internal/provider/data_source_certificate_test.go

```diff
@@ -33,8 +35,160 @@ func TestAccDataSourceCertificate_CertificateContent(t *testing.T) {
 					resource.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.not_before", "2019-11-08T09:01:36Z"),
 					resource.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.not_after", "2019-11-08T19:01:36Z"),
 					resource.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.sha1_fingerprint", "61b65624427d75b61169100836904e44364df817"),
-					testCheckPEMFormat("data.tls_certificate.test", "certificates.0.cert_pem", PreambleCertificate),
-					resource.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.cert_pem", strings.TrimSpace(testTlsDataSourceCertFromContent)+"\n"),
+					tu.TestCheckPEMFormat("data.tls_certificate.test", "certificates.0.cert_pem", PreambleCertificate.String()),
+					resource.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.cert_pem", strings.TrimSpace(fixtures.TestTlsDataSourceCertFromContent)+"\n"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccDataSourceCertificate_UpgradeFromVersion3_4_0(t *testing.T) {
+	resource.UnitTest(t, resource.TestCase{
+
+		Steps: []resource.TestStep{
+			{
+				ExternalProviders: providerVersion340(),
+				Config: `
+					data "tls_certificate" "test" {
+					  content = file("fixtures/certificate.pem")
+					}
+				`,
+				Check: resource.ComposeAggregateTestCheckFunc(
+					resource.TestCheckResourceAttr("data.tls_certificate.test", "certificates.#", "1"),
+
+					resource.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.signature_algorithm", "SHA256-RSA"),
+					resource.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.public_key_algorithm", "RSA"),
+					resource.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.serial_number", "266244246501122064554217434340898012243"),
+					resource.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.is_ca", "false"),
+					resource.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.version", "3"),
+					resource.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.issuer", "CN=Root CA,O=Test Org,L=Here"),
+					resource.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.subject", "CN=Child Cert,O=Child Co.,L=Everywhere"),
+					resource.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.not_before", "2019-11-08T09:01:36Z"),
+					resource.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.not_after", "2019-11-08T19:01:36Z"),
+					resource.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.sha1_fingerprint", "61b65624427d75b61169100836904e44364df817"),
+					tu.TestCheckPEMFormat("data.tls_certificate.test", "certificates.0.cert_pem", PreambleCertificate.String()),
+					resource.TestCheckResourceAttr("data.tls_certificate.test", "certificates.0.cert_pem", strings.TrimSpace(fixtures.TestTlsDataSourceCertFromContent)+"\n"),
+				),
+			},
+			{
```

Comment:
Mmm, most resources are ok but I found this:
```
 resource_cert_request_test.go:143: Step 2/3 error: After applying this test step, the plan was not empty.
        stdout:
        
        
        Terraform used the selected providers to generate the following execution
        plan. Resource actions are indicated with the following symbols:
        -/+ destroy and then create replacement
        
        Terraform will perform the following actions:
        
          # tls_cert_request.test1 must be replaced
        -/+ resource "tls_cert_request" "test1" {
              ~ cert_request_pem = <<-EOT
                    -----BEGIN CERTIFICATE REQUEST-----
                    MIIB3zCCAZECAQAwgcUxCzAJBgNVBAYTAlVTMQswCQYDVQQIEwJDQTEWMBQGA1UE
                    BxMNUGlyYXRlIEhhcmJvcjEZMBcGA1UECRMQNTg3OSBDb3R0b24gTGluazETMBEG
                    A1UEERMKOTU1NTktMTIyNzEVMBMGA1UEChMMRXhhbXBsZSwgSW5jMSgwJgYDVQQL
                    Ex9EZXBhcnRtZW50IG9mIFRlcnJhZm9ybSBUZXN0aW5nMRQwEgYDVQQDEwtleGFt
                    cGxlLmNvbTEKMAgGA1UEBRMBMjAqMAUGAytlcAMhAEMhnupmGkKjFBH7Qey4rEih
                    aBOz8wtD58LjzDty/SFfoIGXMIGUBgkqhkiG9w0BCQ4xgYYwgYMwgYAGA1UdEQR5
                    MHeCC2V4YW1wbGUuY29tggtleGFtcGxlLm5ldIcEfwAAAYcEfwAAAoYmc3BpZmZl
                    Oi8vZXhhbXBsZS10cnVzdC1kb21haW4vd29ya2xvYWSGJ3NwaWZmZTovL2V4YW1w
                    bGUtdHJ1c3QtZG9tYWluL3dvcmtsb2FkMjAFBgMrZXADQQBha457BTY6ioMNdRdt
                    AsAwor2wAW2gFZfRYskAiJX2gGbOl0dqXumoIqgUCkHdKbkIAufCpWMf+Xrr7Gy0
                    0vkP
                    -----END CERTIFICATE REQUEST-----
                EOT -> (known after apply)
              ~ id               = "3930e090e0fc6001930935855837cc3624b16c99" -> (known after apply)
              ~ key_algorithm    = "ED25519" -> (known after apply)
              ~ private_key_pem  = (sensitive value) # forces replacement
                # (3 unchanged attributes hidden)
        
                # (1 unchanged block hidden)
            }
        
        Plan: 1 to add, 0 to change, 1 to destroy.
```

And I believe it's happening because `private_key_pem` is changing from an hash stored via `StateFunc`, to it's plain value.

This is on purpose, but it's also a problem: I have no idea how to address it :(

Reply:
It took 2 good sessions of butting heads with @bendbennett but I think we cracked it.

In essence, I needed to detect the instance when values, currently stored in the state as hashing, are now stored _as-is_, and allow for an _update-in-place_.